bgpd: Metric not set with default route

iqras23 opened the pull request:
Description:

Change is intended for fixing the below listed issue related to default-originate:

- When default route is originated using the neighbor default-originate cmd,
   MED is not set as part of the update message attribute.

Please check the commit for more details.

Co-authored-by: Abhinay Ramesh <rabhinay@vmware.com>
Signed-off-by: Iqra Siddiqui <imujeebsiddi@vmware.com>

ton31337 commented on bgpd/bgpd.c:
```diff
@@ -5106,16 +5106,34 @@ int peer_default_originate_set(struct peer *peer, afi_t afi, safi_t safi,
+			 * to trigger the flow of sending update message.
+			 */
+			if (subgrp &&
+			    CHECK_FLAG(subgrp->sflags,
```
This check_flag looks like NOOP?

iqras23 replied:
Ok, will remove this CHECK_FLAG from both the places.

iqras23 commented on bgpd/bgp_attr.c:
```diff
@@ -965,6 +965,8 @@ struct attr *bgp_attr_default_set(struct attr *attr, uint8_t origin)
 	attr->label = MPLS_INVALID_LABEL;
 	attr->flag |= ATTR_FLAG_BIT(BGP_ATTR_NEXT_HOP);
 	attr->mp_nexthop_len = IPV6_MAX_BYTELEN;
+	attr->med = 0;
```
Default route is originated without Metric.
Hence, setting it up here.

ton31337 replied:
But this is an optional attribute. What is the use case?

iqras23 replied:
Ok, will fix it. 
Thank you :)

iqras23 commented on bgpd/bgp_vty.c:
```diff
@@ -6594,9 +6594,16 @@ DEFUN (no_neighbor_default_originate,
-					      bgp_node_afi(vty),
-					      bgp_node_safi(vty), NULL, 0);
+
+	if (argc == 4)
```
Yes.
So, previously on issuing "no neighbor default originate route-map <WORD>" cmd,
the whole cmd used to vanish away even the default originate config.
With this fix, on issuing no neighbor cmd,
only the route-map application on default originate will be removed,
and default originate config will persist.

iqras23 replied:
For the same reason, we are commenting out those testcases.

ton31337 replied:
What happens if I enter `no neighbor default originate route-map WORD`? argc will be 5, and it's the same as argc 4?

iqras23 commented on tests/bgpd/test_peer_attr.py:
```diff
@@ -84,10 +84,10 @@ class TestFlag(frrtest.TestMultiOut):
-TestFlag.okfail("peer\\ipv4-multicast\\default-originate route-map")
-TestFlag.okfail("peer\\ipv6-unicast\\default-originate route-map")
-TestFlag.okfail("peer\\ipv6-multicast\\default-originate route-map")
+# TestFlag.okfail("peer\\ipv4-unicast\\default-originate route-map")
```
This is because previously on issuing "no neighbor default-originate route-map <WORD>",
the whole config was getting removed. And the testcase was checking for this removal only.
But with the fix, we are keeping the default-originate config intact, so these testcases are no longer valid.
Hence, commenting those out.

Is there a more appropriate way to handle these testcases?
If yes, could you please let us know/help us out.
Thank you :)

ton31337 replied:
test_peer_attr.c has also the same definitions, we might fix all of them maybe?

iqras23 replied:
Hi Donatas,
Based on Igor's comment,
we might have to drop this specific commit.
So, the changes in test_peer_attr will no longer be required.
Thank you :)

iqras23 commented:
> Tests?

Sure.
Will add the relevant tests soon to the PR.

Thank you!

idryzhov reviewed (changes requested):
This change is wrong.

If I use `neighbor X default-originate route-map Y` and then `no neighbor X default-originate route-map Y` then I expect the command to be removed. I don't expect it be half-removed.

If you only want to remove the route-map part, you can simple use `neighbor X default-originate` without setting the route-map.

iqras23 commented:
> 

Hi Igor,
I feel we should always have a no of the optional parameter for every CLI.
We can always remove it if not required.
Please let us know.
Thank you :)

idryzhov commented:
Yes, and you have such ability – reissue the same command without passing optional parameters that you don't need. Most commands in FRR work that way and if some commands work differently then they should be fixed.

The `no` version of any command should always completely delete it from the config.

iqras23 commented:
> Yes, and you have such ability – reissue the same command without passing optional parameters that you don't need. Most commands in FRR work that way and if some commands work differently then they should be fixed.
> 
> The `no` version of any command must always completely delete it from the config.

Ok.
Thanks, Igor.
We will drop this specific commit.

iqras23 commented:
Hi @ton31337, @idryzhov,

We have dropped that CLI specific change.
Also, we have addressed the comment related to the MED commit.
Could you please review again?

Thank you!

ton31337 commented:
ci:rerun failed test_ospfv3_single_area

iqras23 commented:
ci:rerun failed due to compilation issue

iqras23 commented:
ci:rerun failures not related (failed test_multicast_pim_sm_topo3)

iqras23 commented:
Hi @ton31337,

CI has finally passed.
Can we please get this merged?

Thank you :)
